### PR TITLE
setting up SPI for deflate compression support and refactoring of the deflate compression machinery

### PR DESCRIPTION
**Pull request by dromagnoli**
Fixes for
#261 
#262

### CODE REVIEW

File: library/geocore/src/main/java/it/geosolutions/imageio/compression/Compressor.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+     * indicates that compression should end with the current contents of the input buffer.
+     * Some implementations may need it.
+     */
+    void finish();
+
+    /**
+     * compress the input source data section selected by srcOffset and srcLength and
+     * store the on destData, given destOffset and destLength.
+     */
+    int compress(byte[] destData, int srcOffset, int srcLength, int destOffset, int destlength);
```
Comment (aaime) [inline library/geocore/src/main/java/it/geosolutions/imageio/compression/Compressor.java:53]: Thinking out loud, why provide the srcData as a property, but the destdata as a parameter?
Would it make sense to have both as params?

File: library/geocore/src/main/java/it/geosolutions/imageio/compression/Decompressor.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+public interface Decompressor {
+
+    /** Set the input source data buffer to be decompressed */
+    void setInput(byte [] srcData);
+
+    /**
+     * Decompress the source data and put result in the dest data buffer, starting from offset
+     * position, with a maximum number of uncompressed bytes
+     */
+    void decompress(byte[] destData, int offset, int maxUncompressedSize) throws DataFormatException;
```
Comment (aaime) [inline library/geocore/src/main/java/it/geosolutions/imageio/compression/Decompressor.java:50]: Same note about srcData vs destData as in compressor, plus noting that the method signatures are not similar (again, thinking out loud, one might think they should be similar in design, but maybe there is a reason for them to differ).

File: plugin/tiff/src/test/java/it/geosolutions/imageio/tiff/TIFFWriteTest.java [hunk 3]
```diff
@@ -263,17 +263,13 @@ public void imageComplete(ImageWriter source) {
         reader.dispose();
     }
 
-
-
     @Test
     public void writeZSTD() throws IOException {
-
```
Comment (aaime) [inline plugin/tiff/src/test/java/it/geosolutions/imageio/tiff/TIFFWriteTest.java:270]: Only whitespace changes in this file, it seems.

File: plugin/libdeflate/src/main/java/it/geosolutions/imageio/compression/libdeflate/LibDeflateCompressorSpi.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
+package it.geosolutions.imageio.compression.libdeflate;
```
Comment (aaime) [inline plugin/libdeflate/src/main/java/it/geosolutions/imageio/compression/libdeflate/LibDeflateCompressorSpi.java:1]: No copyright header.

File: plugin/libdeflate/src/main/java/it/geosolutions/imageio/compression/libdeflate/LibDeflateDecompressor.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package it.geosolutions.imageio.compression.libdeflate;
```
Comment (aaime) [inline plugin/libdeflate/src/main/java/it/geosolutions/imageio/compression/libdeflate/LibDeflateDecompressor.java:1]: No copyright header.

File: plugin/libdeflate/src/main/java/it/geosolutions/imageio/compression/libdeflate/LibDeflateDecompressorSpi.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
+package it.geosolutions.imageio.compression.libdeflate;
```
Comment (aaime) [inline plugin/libdeflate/src/main/java/it/geosolutions/imageio/compression/libdeflate/LibDeflateDecompressorSpi.java:1]: No copyright header.

**Review by aaime — COMMENTED** [review]
I don't see tests for the integration of libdeflate (might be tricky, needs to be guarded on the availability of the native lib, but still, something should be there).

**Comment by dromagnoli** [comment]
> I don't see tests for the integration of libdeflate (might be tricky, needs to be guarded on the availability of the native lib, but still, something should be there).

Sure. It's still part of the WIP. I need to deploy the needed dependency yet

**Comment by dromagnoli** [comment]
> I don't see tests for the integration of libdeflate (might be tricky, needs to be guarded on the availability of the native lib, but still, something should be there).

Sure, It's still part of the WIP. I need to deploy the needed dependency yet.

File: library/geocore/src/main/java/it/geosolutions/imageio/compression/Compressor.java (hunk 1 above)
Comment (dromagnoli) [reply library/geocore/src/main/java/it/geosolutions/imageio/compression/Compressor.java:53]: the libdeflate implementation may need to call finish before doing the compression and it needed to get the source data first.

File: library/geocore/src/main/java/it/geosolutions/imageio/compression/Decompressor.java (hunk 2 above)
Comment (dromagnoli) [reply library/geocore/src/main/java/it/geosolutions/imageio/compression/Decompressor.java:50]: Both implementations of the decompressors simply decompress whatever has been provided as source.

File: plugin/tiff/src/main/java/it/geosolutions/imageio/compression/zipdeflate/ZipDeflateCompressorSpi.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+public class ZipDeflateCompressorSpi extends AbstractCompressorSpi implements CompressorSpi {
+
+    public static Set<CompressionType> SUPPORTED_TYPES = Collections.singleton(CompressionType.DEFLATE);
+
+    public ZipDeflateCompressorSpi() {
+        super();
+        setMaxLevel(10);
+        setMinLevel(0);
+        setPriority(2);
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply plugin/tiff/src/main/java/it/geosolutions/imageio/compression/zipdeflate/ZipDeflateCompressorSpi.java:19]: my fault. I was doing some tests and forgot to remove it.
good catch

File: plugin/tiff/src/main/java/it/geosolutions/imageioimpl/plugins/tiff/TIFFDeflater.java [hunk 8]
```diff
@@ -141,30 +145,33 @@ public int encode(byte[] b, int off,
...
         } else {
-            deflater.setInput(b, off, height*scanlineStride);
-            deflater.finish();
-
-            numCompressedBytes = deflater.deflate(compData);
+            deflateCompressor.setInput(b);
+            deflateCompressor.finish();
+            numCompressedBytes =
+                    deflateCompressor.compress(compData, off, height*scanlineStride, 0, compData.length);
+            /*if (!useLibDeflate) {
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply plugin/tiff/src/main/java/it/geosolutions/imageioimpl/plugins/tiff/TIFFDeflater.java:166]: Good catch. Forgot to cleanup. thanks

File: plugin/tiff/src/test/java/it/geosolutions/imageio/tiff/TIFFWriteTest.java (hunk 3 above)
Comment (dromagnoli) [reply plugin/tiff/src/test/java/it/geosolutions/imageio/tiff/TIFFWriteTest.java:270]: Ops... a rebase issue, I think. fixing

File: plugin/tiff/src/main/java/META-INF/MANIFEST.MF [hunk 9]
```diff
@@ -0,0 +1,10 @@
+Manifest-Version: 1.0
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply plugin/tiff/src/main/java/META-INF/MANIFEST.MF:1]: Not idea on how it ended there. removing it

File: pom.xml [hunk 10]
```diff
@@ -431,7 +432,11 @@
       <artifactId>imageio-ext-utilities</artifactId>
       <version>${project.version}</version>
      </dependency>
-
+     <dependency>
+      <groupId>it.geosolutions.imageio-ext</groupId>
+      <artifactId>imageio-ext-libdeflate-java</artifactId>
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply pom.xml:437]: actually that was a temporary "workaround". it contains the jar of the [libdeflate-java](https://github.com/astei/libdeflate-java) lib that I have built so far.
I have contacted the author of the lib about plan for deployment on maven and/or a beta release or whatever but I didn't get an answer so far.
we might want to deploy it on our geosolutions repository, using the suggested [dependency](https://github.com/astei/libdeflate-java#maven)

<dependency>
    <groupId>me.steinborn</groupId>
    <artifactId>libdeflate-java-core</artifactId>
    <version>0.1.0-SNAPSHOT</version>
</dependency>

**Comment by dromagnoli** [comment]
> > I don't see tests for the integration of libdeflate (might be tricky, needs to be guarded on the availability of the native lib, but still, something should be there).
> 
> Sure, It's still part of the WIP. I need to deploy the needed dependency yet.

Tests Added

**Pull request merged**